pypesto/objective/history.py

```diff
@@ -882,89 +884,100 @@ def update(self,
                result: ResultType) -> None:
         """Update history and best found value."""
         self.history.update(x, sensi_orders, mode, result)
-        self._update_vals(x, sensi_orders, mode, result)
+        self._update_vals(x, result)
 
     def finalize(self):
         self.history.finalize()
 
     def _update_vals(self,
                      x: np.ndarray,
-                     sensi_orders: Tuple[int],
-                     mode: str,
                      result: ResultType):
         """
         Update initial and best function values.
         """
         # update initial point
-        if self.fval0 is None and np.allclose(x, self.x0) \
-                and 0 in sensi_orders:
-            if mode == MODE_FUN:
-                self.fval0 = result[FVAL]
-                self.x0 = x
-            else:  # mode == MODE_RES:
-                chi2 = res_to_chi2(result[RES])
-                self.fval0 = result.get(FVAL, chi2)
-                self.x0 = x
+        if np.allclose(x, self.x0):
+            if self.fval0 is None:
+                self.fval0 = result.get(FVAL, None)
+
+            self.x0 = x
 
         # update best point
-        if 0 in sensi_orders:
-            # extract function value
-            if mode == MODE_FUN:
-                fval = result[FVAL]
-            else:  # mode == MODE_RES:
-                chi2 = res_to_chi2(result[RES])
-                fval = result.get(FVAL, chi2)
-            # store value
-            if fval < self.fval_min:
-                self.fval_min = fval
-                self.x_min = x
-                self.grad_min = result.get(GRAD)
-                self.hess_min = result.get(HESS)
-                self.res_min = result.get(RES)
-                self.sres_min = result.get(SRES)
+        fval = result.get(FVAL, None)
+        grad = result.get(GRAD, None)
+        hess = result.get(HESS, None)
+        res = result.get(RES, None)
+        sres = result.get(SRES, None)
+        if fval is not None and fval < self.fval_min:
+            self.fval_min = fval
+            self.x_min = x
+            self.grad_min = grad
+            self.hess_min = hess
+            self.res_min = res
+            self.sres_min = sres
+
+        x_same = False
```

Comment:
can we just have `x_same = np.all(self.x_min == x)`` minimum (?) computational effort, better readable

Reply:
and add a comment that the following is due to gradients often being evaluated only after a good point has been found, in subsequent iterations?